hilti/toolchain/include/ast/type.h

```diff
@@ -2,414 +2,582 @@
-class isSortable {};
-class isDereferenceable {};
-class isIterable {};
-class isIterator {};
-class isMutable {};
-class isParameterized {};
-class isReferenceType {};
-class isRuntimeNonTrivial {};
-class isView {};
-class isViewable {};
-class supportsWildcard {};
-class takesArguments {};
-} // namespace trait
+/** Strong type argument to `Unification` constructor. */
+struct NeverMatch {
+    explicit NeverMatch() = default;
+};
 
+namespace detail {
 using ResolvedState = std::unordered_set<uintptr_t>;
+}
+
+} // namespace type
+
+namespace type {
+
+/**
+ * Represent a type's unification string. Two types with the same unification
+ * string are considered equivalent during AST processing.
+ */
+struct Unification {
+    /** Creates a unset unification string, which will never match any other. */
+    Unification() = default;
+
+    /**
+     * Create a unification from a pre-computed serialization string.
+     *
+     * @param serialization string representation of the unification; must not be empty
+     */
+    Unification(std::string serialization) : _serialization(std::move(serialization)) {
+        assert(! _serialization->empty());
+    }
 
-/** Additional flags to associated with types. */
-enum class Flag {
-    /** Set to make the type `const`. */
-    Constant = (1U << 0U),
+    /**
+     * Create a unification that's guaranteed to never match any other unification.
+     *
+     * @param never_match unused
+     */
+    Unification(NeverMatch _) : _serialization("") {}
 
-    /** Set to make the type `non-const`. */
-    NonConstant = (1U << 1U),
+    Unification(const Unification& other) = default;
+    Unification(Unification&& other) = default;
+    ~Unification() = default;
+
+    Unification& operator=(const Unification& other) = default;
+    Unification& operator=(Unification&& other) = default;
 
     /**
-     * Marks the type as having a top-level scope that does not derive scope content
-     * from other nodes above it in the AST (except for truly global IDs).
+     * Returns a string representation of the unification string. This is
+     * human-readable for display purposes.
      */
-    NoInheritScope = (1U << 2U),
+    std::string str() const {
+        if ( ! _serialization.has_value() )
+            return "<unset>";
+
+        if ( _serialization->empty() )
+            return "<never-match>";
+
+        return *_serialization;
+    }
+
+    /** Forwards to `str()`. */
+    operator std::string() const { return str(); }
 
-    /** When walking over an AST, skip this node's children. This allows to
-     * break cycles. */
-    PruneWalk = (1U << 3U),
+    /** Returns true if unification string has been set. */
+    operator bool() const { return _serialization.has_value(); }
+
+    /**
+     * Returns true if two unifications are equivalent. Will always return
+     * false if any of the is set to never-match, or not set at all.
+     */
+    bool operator==(const Unification& other) const {
```

Comment:
Nit but since this is new code: a subjectively better way to implement these operators is to implement `friend bool operator==(const Unification&, const Unification&)`. This pattern enables calling implicit conversions to `Unification` symmetrically for both the LHS and RHS while the non-`friend` version only allows this for the RHS.

Same for `operator!=`.

https://google.github.io/styleguide/cppguide.html#:~:text=For%20a%20type,and%20ordering%20operators.